vault/logical_system_namespaces_test.go

```diff
@@ -0,0 +1,230 @@
+		require.NoError(t, err)
+		outputMetadata := res.Data["custom_metadata"].(map[string]string)
+		require.Equal(t, outputMetadata, map[string]string{"testing": "hello"},
+			"expected old key to be removed and new key to be added")
+	})
+}
+
+func TestNamespaceBackend_Delete(t *testing.T) {
+	t.Parallel()
+	b := testSystemBackend(t)
+
+	t.Run("delete existing namespace", func(t *testing.T) {
+		testCreateNamespace(t, b, "foo", nil)
+
+		req := logical.TestRequest(t, logical.DeleteOperation, "namespaces/foo")
+		_, err := b.HandleRequest(context.Background(), req)
+		require.NoError(t, err)
+
+		req = logical.TestRequest(t, logical.ReadOperation, "namespaces/foo")
+		res, err := b.HandleRequest(context.Background(), req)
+		require.NoError(t, err)
+		require.Empty(t, res, "expected emtpy response to read after deleting namespace")
+	})
+
+	t.Run("delete non-existent namespace", func(t *testing.T) {
+		req := logical.TestRequest(t, logical.DeleteOperation, "namespaces/bar")
+		_, err := b.HandleRequest(context.Background(), req)
+		require.NoError(t, err)
+	})
+}
+
+func TestNamespaceBackend_List(t *testing.T) {
+	b := testSystemBackend(t)
+
+	t.Run("list is empty if root is only namespace", func(t *testing.T) {
+		req := logical.TestRequest(t, logical.ListOperation, "namespaces")
+		res, err := b.HandleRequest(context.Background(), req)
+		require.NoError(t, err)
+		require.Equal(t, res.Data, map[string]interface{}{}, "list data has unexpected elements")
+	})
+
+	t.Run("list includes non-root namespaces", func(t *testing.T) {
+		testCreateNamespace(t, b, "foo", nil)
+		testCreateNamespace(t, b, "bar", nil)
+
+		req := logical.TestRequest(t, logical.ListOperation, "namespaces")
+		res, err := b.HandleRequest(context.Background(), req)
+
+		require.NoError(t, err)
+		require.NotEmpty(t, res.Data["keys"], "keys is empty")
+		require.NotEmpty(t, res.Data["key_info"], "key_info is empty")
+
+		keys, ok := res.Data["keys"].([]string)
+		require.True(t, ok, "keys is not a list")
+		keyInfo, ok := res.Data["key_info"].(map[string]interface{})
+		require.True(t, ok, "key_info is not a map")
+
+		require.Equal(t, len(keys), 2, "expected two entries in keys")
+		require.Equal(t, len(keyInfo), 2, "expected two entries in key_info")
+
+		for _, path := range keys {
+			info, ok := keyInfo[path]
+			require.True(t, ok, fmt.Sprintf("key_info does not have path %q which is present in keys", path))
+			var ns namespace.Namespace
+			require.NoError(t, mapstructure.Decode(info, &ns), "key_info entry is not a namespace")
+			require.Equal(t, ns.Path, path, "path in key does not match path in namespace struct")
+			require.NotEmpty(t, ns.ID, "namespace ID should not be empty")
+			require.NotEqual(t, ns.ID, namespace.RootNamespaceID, "list should not include root namespace")
+		}
+	})
+}
+
+func TestNamespaceBackend_Scan(t *testing.T) {
+	// TODO(satoqz): Implement scan once a clear API is available from NamespaceStore.
```

Comment:
Wouldn't scan just list all? Regardless, I think this can be a follow-up ticket.